p2p/libp2p/connectionMonitor/libp2pConnectionMonitor2.go

```diff
@@ -0,0 +1,124 @@
+	thresholdMinConnectedPeers int
+	sharder                    Sharder
+}
+
+// NewLibp2pConnectionMonitor2 creates a new connection monitor (version 2 that is more streamlined and does not care
+//about pausing and resuming the discovery process)
+func NewLibp2pConnectionMonitor2(reconnecter p2p.Reconnecter, thresholdMinConnectedPeers int) (*libp2pConnectionMonitor2, error) {
+	if thresholdMinConnectedPeers < 0 {
+		return nil, p2p.ErrInvalidValue
+	}
+
+	cm := &libp2pConnectionMonitor2{
+		reconnecter:                reconnecter,
+		chDoReconnect:              make(chan struct{}),
+		thresholdMinConnectedPeers: thresholdMinConnectedPeers,
+	}
+
+	if reconnecter != nil {
+		go cm.doReconnection()
+	}
+
+	return cm, nil
+}
+
+// Listen is called when network starts listening on an addr
+func (lcm *libp2pConnectionMonitor2) Listen(network.Network, multiaddr.Multiaddr) {}
+
+// ListenClose is called when network stops listening on an addr
+func (lcm *libp2pConnectionMonitor2) ListenClose(network.Network, multiaddr.Multiaddr) {}
+
+// Request a reconnect to initial list
+func (lcm *libp2pConnectionMonitor2) doReconn() {
+	select {
+	case lcm.chDoReconnect <- struct{}{}:
+	default:
+	}
+}
+
+// Connected is called when a connection opened
+func (lcm *libp2pConnectionMonitor2) Connected(netw network.Network, conn network.Conn) {
+	if lcm.sharder != nil {
+		evicted := lcm.sharder.ComputeEvictList(conn.RemotePeer(), netw.Peers())
+		for _, pid := range evicted {
+			_ = netw.ClosePeer(pid)
+		}
+	}
+}
+
+// Disconnected is called when a connection closed
+func (lcm *libp2pConnectionMonitor2) Disconnected(netw network.Network, _ network.Conn) {
+	lcm.doReconnectionIfNeeded(netw)
+}
+
+func (lcm *libp2pConnectionMonitor2) doReconnectionIfNeeded(netw network.Network) {
+	if !lcm.IsConnectedToTheNetwork(netw) {
+		lcm.doReconn()
+	}
+}
+
+// OpenedStream is called when a stream opened
+func (lcm *libp2pConnectionMonitor2) OpenedStream(network.Network, network.Stream) {}
+
+// ClosedStream is called when a stream closed
+func (lcm *libp2pConnectionMonitor2) ClosedStream(network.Network, network.Stream) {}
+
+func (lcm *libp2pConnectionMonitor2) doReconnection() {
+	for {
+		<-lcm.chDoReconnect
+		<-lcm.reconnecter.ReconnectToNetwork()
+
+		time.Sleep(DurationBetweenReconnectAttempts)
+	}
+}
+
+// IsConnectedToTheNetwork returns true if the number of connected peer is at least equal with thresholdMinConnectedPeers
+func (lcm *libp2pConnectionMonitor2) IsConnectedToTheNetwork(netw network.Network) bool {
+	return len(netw.Peers()) >= lcm.thresholdMinConnectedPeers
+}
+
+// SetThresholdMinConnectedPeers sets the minimum connected peers number when the node is considered connected on the network
+//TODO(iulian) refactor this in a future PR (not require to inject the netw pointer)
+func (lcm *libp2pConnectionMonitor2) SetThresholdMinConnectedPeers(thresholdMinConnectedPeers int, netw network.Network) {
+	if netw == nil {
```

Comment:
check.ifnil?

Reply:
check.IfNilReflect in this case as the interface is defined in libp2p's libraries.